### DISCUSSION TITLE
Some detailed questions about utility calculation in oneshot

### DISCUSSION THREAD

**Discussion by rezunli96** [yasserfarouk/scml#16] [one-shot-qa] [answered]
In Section 2.3 of the oneshot document, to my understanding it adopts a greedy ranking approach to define the satisfiable contract:
<img width="690" alt="Screen Shot 2021-04-06 at 12 13 29" src="https://user-images.githubusercontent.com/29717251/113743337-9231ca80-96d1-11eb-9c99-142fc2ca0ae3.png">

First question: why we have to define it in this way? Intuitively for the satisfiable input contract it could also be argmin (to make the agent spend as less as possible)?

Next part I am a little bit confused:

<img width="666" alt="Screen Shot 2021-04-06 at 12 17 30" src="https://user-images.githubusercontent.com/29717251/113743827-17b57a80-96d2-11eb-9c62-e1c0c6bfcf6b.png">

It appears that in terms of calculating Q^{i+}_a, Q^{o+}_a only C^i and C^o matters. And only \hat{C}^o appears qt equation (1) when calculating the sale profit. So what's the essential functionality of the greedy ranking in profit computation?

I check the code where in oneshot/world.py
```    
def on_contract_signed(self, contract: Contract) -> bool:
        product = contract.annotation["product"]
        bought = contract.agreement["quantity"]
        buy_cost = bought * contract.agreement["unit_price"]
        oldq = self._sold_quantity[product, self.current_step + 1]
        oldp = self._real_price[product, self.current_step + 1]
        totalp = 0.0
        if oldq > 0:
            totalp = oldp * oldq
        self._sold_quantity[product, self.current_step + 1] += bought
        self._real_price[product, self.current_step + 1] = (totalp + buy_cost) / (
            oldq + bought
        )
        self._input_quantity[contract.annotation["buyer"]] += bought
        self._input_price[contract.annotation["buyer"]] += buy_cost
        self._output_quantity[contract.annotation["seller"]] += bought
        self._output_price[contract.annotation["seller"]] += buy_cost
        contract.executed_at = self.current_step
        return super().on_contract_signed(contract)
```
It seems it does not adopt the greedy approach as written in the document but just directly add all the quantities?

In oneshot/ufun.py I guess it is the only place where the greedy ranking is implemented:

```
    def from_offers(self, offers: Iterable[Tuple], outputs: Iterable[bool]) -> float:
        """
        Calculates the utility value given a list of offers and whether each offer is for output or not.
        """
        qin, qout, pin, pout = 0, 0, 0, 0
        output_offers = []
        for offer, is_output in zip(offers, outputs):
            offer = self.outcome_as_tuple(offer)
            if is_output:
                output_offers.append(offer)
            else:
                qin += offer[QUANTITY]
                pin += offer[UNIT_PRICE] * offer[QUANTITY]
        n_lines = self.n_lines
        output_offers = sorted(output_offers, key=lambda x: -x[UNIT_PRICE])
        producible = min(qin, n_lines)
        for offer in output_offers:
            qout += offer[QUANTITY]
            if qout >= producible:
                qout = producible
                break
        return self.from_aggregates(qin, qout, pin, pout)
```

But it appears there only the \hat{C}^o is calculated. So do we actually need not to calculate \hat{C}^i as described in the document?

**Comment by yasserfarouk** [comment, marked as answer]
Hi, great questions.

``` why we have to define it in this way? Intuitively for the satisfiable input contract it could also be argmin (to make the agent spend as less as possible)?```
Agents are **required** to satisfy all their contracts. They do not have a choice. If we assume that the agen's balance is infinite, then there is actually no need to solve this argmax at all and \bar C^i will exactly equal C^i. But agents have finite balance that may not be enough to pay for C^i. That is why we allow them to pick which contracts to satisfy **at the day they are going bankrupt**. Nevertheless, they are still forced to exhaust all of their money before becoming bankrupt. So the argmax. If an argmin is used, it would have led to \bar C^i = \phi (no trade) which is simply not an option for the agent.

```It appears that in terms of calculating Q^{i+}_a, Q^{o+}_a only C^i and C^o matters. And only \hat{C}^o appears qt equation (1) when calculating the sale profit. So what's the essential functionality of the greedy ranking in profit computation?```
I am not 100% sure that I understand the question but I will try my best here:

- Q^{i+}_a is the extra input the agent bought for which they will have to pay disposable cost. We actually over-penalize the agent by having it pay disposal cost for **EVERYTHING it signed to buy** not only for the amount it actually could pay for (by summing over Q^i instead of \bar Q^i). This is only relevant when \bar Q^i is not equal to Q^i and this will happen only **at the day the agent becomes bankrupt** so you can think of this extra payment as a bankruptcy **penalty**. 
- Q^{o+}_a is the quantity of sales the agent could not satisfy. It is what it pays shortfall penalty for. Here it is more straight forward. Just the difference between what it should have sold and what it could produce (because it sells everything it produces).
- There are two greedy rankings of contracts we used:
  - Ascending by price when calculating \bar Q^{i+} to allow the agent to buy as much as it can from the amount it has binding contracts to buy
  - Descending by price to give priority to sell contracts with higher prices in order to increase its total income when there is a shortfall.

```It seems it does not adopt the greedy approach as written in the document but just directly add all the quantities?```
These calculations are used internally by the system to calculate trading prices and it is orthogonal to fun calculation.
The relevant code can be found in ```from_offers``` method of ```OneShotUfun```  [here](https://github.com/yasserfarouk/scml/blob/master/src/scml/oneshot/ufun.py)


```But it appears there only the \hat{C}^o is calculated. So do we actually need not to calculate \hat{C}^i as described in the document?
```

Actually that is an older version of the code that did not implement this ranking. Sorry for that. This bug was solved already and the relevant code from  [ufun.py](https://github.com/yasserfarouk/scml/blob/master/src/scml/oneshot/ufun.py) is this:

```python
 def order(x):
            """A helper function to order contracts in the following fashion:
            1. input contracts are ordered from cheapest to most expensive.
            2. output contracts are ordered from highest price to cheapest.
            3. The relative order of input and output contracts is indeterminate.
            """
            offer, is_output, is_exogenous = x
            # if is_exogenous and self.force_exogenous:
            #     return float("-inf")
            return -offer[UNIT_PRICE] if is_output else offer[UNIT_PRICE]
```

```python
# sort contracts in the optimal order of execution: from cheapest when
# buying and from the most expensive when selling. See `order` above.
sorted_offers = list(sorted(zip(offers, outputs, exogenous), key=order))
```

Finally, this greedy ordering of input contracts will **ONLY** be relevant at the day the agent becomes bankrupt (if it ever does) and it will have no other effect any other time. Becoming bankrupt is so bad that the small effect of the final balance resulting from the ordering is not that important. The main reason we fix this step of processing is to be sure that the ufun calculation is deterministic.

thanks

**Reply by rezunli96** [reply]
Thanks it makes clear to me now. Appreciate your detailed explanation and code!

**Comment by rezunli96** [comment]
Hello, just some follow-up questions. In oneshot/ufun.py, the from_offer() method:
```
        for offer, is_output, is_exogenous in sorted_offers:
            offer = self.outcome_as_tuple(offer)
            if is_output:
                output_offers.append((offer, is_exogenous))
                continue
            topay_this_time = offer[UNIT_PRICE] * offer[QUANTITY]
            if not going_bankrupt and (
                pin + topay_this_time + offer[QUANTITY] * self.production_cost
                > self.current_balance
            ):
                unit_total_cost = offer[UNIT_PRICE] + self.production_cost
                can_buy = int((self.current_balance - pin) // unit_total_cost)
                qin_bar = qin + can_buy
                going_bankrupt = True
            pin += topay_this_time
            qin += offer[QUANTITY]
```

It looks unclear to me why we should involve production_cost of the current buy offer being considered, to decide whether the agent bankrupt or not and calculate can_buy. I think it either should be that (1) do not consider any production cost, so an agent can be declared bankrupt only after execute buy contract but before any production or (2) consider the production costs of every previously executed contracted, so it is like goods get manufactured immediately before it proceeds to the next offer. The current implementation is like it assume the current offer being considered will be manufactured first.

And I think here going_bankrupt doesn't suggest the agent would actually bankrupt right? It is just an auxiliary variable to calculate the producibility. And it could happen the agent get compensated from sale profit so overall it wouldn't get bankrupted?

**Reply by yasserfarouk** [reply]
> It looks unclear to me why we should involve production_cost of the current buy offer being considered, to decide whether the agent bankrupt or not and calculate can_buy.

What this calculation is doing is just calculating a limit of the number of items that can be produced by the agent. It is not really evaluating whether the agent is going bankrupt or not. As you say later:

> And I think here going_bankrupt doesn't suggest the agent would actually bankrupt 

yes it is just an auxiliary variable.

> And it could happen the agent get compensated from sale profit so overall it wouldn't get bankrupted?

yes again. 

Today, I plan to publish an update of the rules document that explicitly states this to avoid any confusion.

**Reply by rezunli96** [reply]
Thank you!

**Reply by yasserfarouk** [reply]
The update is published
